numpyro/distributions/continuous.py

```diff
@@ -3265,6 +3265,301 @@ def infer_shapes(
+            jnp.shape(concentration)[:-2], jnp.shape(self.scale_tril)[:-2]
+        )
+        event_shape = jnp.shape(self.scale_tril)[-2:]
+        self.concentration = concentration[..., 0, 0]
+        super().__init__(
+            batch_shape=batch_shape,
+            event_shape=event_shape,
+            validate_args=validate_args,
+        )
+
+    @validate_sample
+    def log_prob(self, value: ArrayLike) -> ArrayLike:
+        # L = value (Cholesky factor), X = L @ L^T ~ InverseWishart(Psi, nu)
+        # log p(X) = (nu/2) log|Psi| - (nu*p/2) log(2) - log Gamma_p(nu/2)
+        #            - ((nu+p+1)/2) log|X| - tr(Psi @ X^{-1}) / 2
+        # Trace trick: tr(Psi @ X^{-1}) = ||L^{-1} @ scale_tril||_F^2
+        x = solve_triangular(*jnp.broadcast_arrays(value, self.scale_tril), lower=True)
+        trace = jnp.square(x).sum(axis=(-1, -2))
+
+        p = value.shape[-1]
+        return (
+            self.concentration * tri_logabsdet(self.scale_tril)  # (nu/2) log|Psi|
+            + p * (1 - self.concentration / 2) * jnp.log(2)  # normalization
+            - multigammaln(self.concentration / 2, p)
+            - (self.concentration + p + 1)
+            * tri_logabsdet(value)  # -((nu+p+1)/2) log|X|
+            - trace / 2  # -tr(Psi @ X^{-1}) / 2
+            + jnp.sum(  # Jacobian: X = L @ L^T
+                jnp.arange(p, 0, -1) * jnp.log(jnp.diagonal(value, axis1=-2, axis2=-1)),
+                axis=-1,
+            )
+        )
+
+    @lazy_property
+    def scale_matrix(self):
+        return jnp.matmul(self.scale_tril, self.scale_tril.mT)
+
+    @lazy_property
+    def rate_matrix(self):
+        identity = jnp.broadcast_to(
+            jnp.eye(self.scale_tril.shape[-1]), self.scale_tril.shape
+        )
+        return cho_solve((self.scale_tril, True), identity)
+
+    def sample(
+        self, key: jax.dtypes.prng_key, sample_shape: tuple[int, ...] = ()
+    ) -> ArrayLike:
+        assert is_prng_key(key)
+        # If X ~ Wishart(nu, Psi^{-1}), then X^{-1} ~ InverseWishart(nu, Psi)
+        # Use Bartlett decomposition to sample Wishart, then invert.
+        rng_diag, rng_offdiag = random.split(key)
+        latent = jnp.zeros(sample_shape + self.batch_shape + self.event_shape)
+        p = self.event_shape[-1]
+        i = jnp.arange(p)
+        latent = latent.at[..., i, i].set(
+            jnp.sqrt(
+                random.chisquare(
+                    rng_diag, self.concentration[..., None] - i, latent.shape[:-1]
+                )
+            )
+        )
+        i, j = jnp.tril_indices(p, -1)
+        assert i.size == p * (p - 1) // 2
+        latent = latent.at[..., i, j].set(
+            random.normal(rng_offdiag, latent.shape[:-2] + (i.size,))
+        )
+        # L_wishart @ L_wishart^T ~ Wishart(nu, Psi^{-1})
+        scale_tril_wishart = cholesky_of_inverse(self.scale_matrix)
+        L_wishart = jnp.matmul(*jnp.broadcast_arrays(scale_tril_wishart, latent))
```

Comment:
can we use solve_triangular(self.scale_matrix, latent, lower=True) instead?

Reply:
I tried it out, and I could not make it work with `solve_triangular(self.scale_tril, latent, lower=True)` (or with `.mT`)  as

- **Current approach:** `chol(Psi^{-1}) @ latent` produces a **lower triangular** `L_wishart`, which is required for the subsequent `solve_triangular` inversion step.

- **The suggested approach:** `L^{-T} @ latent` (upper × lower) produces a *general (non-triangular) matrix* (as far as I can see), breaking the `solve_triangular` call that follows.

The triangular structure must be preserved throughout the sampling algorithm, so `cholesky_of_inverse(self.scale_matrix)` is necessary.

(also verified both by hand and with Opus 🙈 )